### ISSUE TITLE
Prioritize which operator list is built on the worker.

### ISSUE THREAD

**Issue by bf** [mozilla/pdf.js#1887]
Hello, 

I have experienced huge lags in PDF rendering when "skimming" (scrolling fast) over a 300+ page PDF file. According to my console, calls to `getOperatorList` take more than 1000ms each which I assume is the core of this performance issue.

It seems that for each page you scroll over, `getOperatorList` is called. The problem is, that this call finishes when the user has scrolled further for about 50 pages, so it is no longer relevant as it returns a result for a page that is not visible to the user any more. The end result is, that when I have reached page 350, I have to way for 10+ seconds until the currently visible document is rendered.

In my eyes, this performance issue can probably be saved by using a LIFO queue for rendering pages, so the last page is rendered as soon as possible, and then all the pages I have scrolled over.

Enclosed you find parts of the console output. Thank you!

```
/pdf.js/build/pdf.js:30363      page=   426 - getOperatorList: time=1189ms, len=1995
l10n.js:194                         [l10n] #thumb_page_canvas missing for [de]
/pdf.js/build/pdf.js:30363      page=   427 - getOperatorList: time=1022ms, len=1553
l10n.js:194                         [l10n] #thumb_page_canvas missing for [de]
/pdf.js/build/pdf.js:30363      page=   428 - getOperatorList: time=1294ms, len=2755
l10n.js:194                         [l10n] #thumb_page_canvas missing for [de]
/pdf.js/build/pdf.js:30363      page=   448 - getOperatorList: time=1324ms, len=87
/pdf.js/build/pdf.js:30363      page=   474 - getOperatorList: time=1027ms, len=1305
/pdf.js/build/pdf.js:30363      page=   484 - getOperatorList: time=1366ms, len=2564
/pdf.js/build/pdf.js:30363      page=   496 - getOperatorList: time=1089ms, len=1110
/pdf.js/build/pdf.js:30363      page=   498 - getOperatorList: time=1171ms, len=1010
/pdf.js/build/pdf.js:30363      page=   499 - getOperatorList: time=1064ms, len=612
/pdf.js/build/pdf.js:30363      page=   517 - getOperatorList: time=1412ms, len=2277
/pdf.js/build/pdf.js:30363      page=   520 - getOperatorList: time=1298ms, len=1958
/pdf.js/build/pdf.js:30363      page=   521 - getOperatorList: time=1264ms, len=2019
/pdf.js/build/pdf.js:30363      page=   522 - getOperatorList: time=1243ms, len=1971
l10n.js:194                         [l10n] #thumb_page_canvas missing for [de]
/pdf.js/build/pdf.js:30363      page=   523 - getOperatorList: time=952ms, len=76
l10n.js:194                         [l10n] #thumb_page_canvas missing for [de]
l10n.js:194                         [l10n] #thumb_page_canvas missing for [de]
l10n.js:194                         [l10n] #thumb_page_canvas missing for [de]
/pdf.js/build/pdf.js:30363      page=   519 - getOperatorList: time=1493ms, len=2647
l10n.js:194                         [l10n] #thumb_page_canvas missing for [de]
/pdf.js/build/pdf.js:30363      page=   508 - getOperatorList: time=1190ms, len=1166
/pdf.js/build/pdf.js:30363      page=   505 - getOperatorList: time=1328ms, len=1135
/pdf.js/build/pdf.js:30363      page=   502 - getOperatorList: time=1263ms, len=47
/pdf.js/build/pdf.js:30363      page=   500 - getOperatorList: time=1157ms, len=1657
/pdf.js/build/pdf.js:30363      page=   489 - getOperatorList: time=1633ms, len=1620
/pdf.js/build/pdf.js:30363      page=   485 - getOperatorList: time=1059ms, len=1046
/pdf.js/build/pdf.js:30363      page=   482 - getOperatorList: time=964ms, len=619
/pdf.js/build/pdf.js:30363      page=   478 - getOperatorList: time=1029ms, len=648
/pdf.js/build/pdf.js:30363      page=   475 - getOperatorList: time=891ms, len=531
/pdf.js/build/pdf.js:30363      page=   473 - getOperatorList: time=1196ms, len=1687
/pdf.js/build/pdf.js:30363      page=   470 - getOperatorList: time=1094ms, len=1314
/pdf.js/build/pdf.js:30363      page=   468 - getOperatorList: time=1215ms, len=1861
/pdf.js/build/pdf.js:30363      page=   465 - getOperatorList: time=1380ms, len=1735
/pdf.js/build/pdf.js:30363      page=   461 - getOperatorList: time=1239ms, len=1859
/pdf.js/build/pdf.js:30363      page=   458 - getOperatorList: time=1136ms, len=1294
/pdf.js/build/pdf.js:30363      page=   455 - getOperatorList: time=1083ms, len=1667
l10n.js:194                         [l10n] #search_terms_not_found missing for [de]
/pdf.js/build/pdf.js:30363      page=   451 - getOperatorList: time=1484ms, len=1722
/pdf.js/build/pdf.js:30363      page=   442 - getOperatorList: time=1052ms, len=703
/pdf.js/build/pdf.js:30363      page=   421 - getOperatorList: time=1336ms, len=1749
/pdf.js/build/pdf.js:30363      page=   404 - getOperatorList: time=1161ms, len=1265
/pdf.js/build/pdf.js:30363      page=   377 - getOperatorList: time=1145ms, len=988
l10n.js:194                         [l10n] #thumb_page_canvas missing for [de]
/pdf.js/build/pdf.js:30363      page=   365 - getOperatorList: time=1375ms, len=2242
/pdf.js/build/pdf.js:30363      page=   356 - getOperatorList: time=1067ms, len=1612
/pdf.js/build/pdf.js:30363      page=   348 - getOperatorList: time=1486ms, len=4459
/pdf.js/build/pdf.js:30363      page=   342 - getOperatorList: time=1408ms, len=22
/pdf.js/build/pdf.js:30363      page=   338 - getOperatorList: time=1320ms, len=1575
/pdf.js/build/pdf.js:30363      page=   335 - getOperatorList: time=1588ms, len=103
```

**Comment by evadne** [comment]
Just to add to #8384 I can also see that in the worker’s `RenderPageRequest` handler (`wphSetupRenderPage`), `task.terminated` is checked once `getOperatorList` completes.

But should we also try to check once more (and earlier), when `getPage` just completed too, so as to avoid doing too much work.

**Comment by yurydelendik** [comment]
Please submit a PR with proposed solution above.

**Comment by timvandermeij** [comment]
Closing since this is (for the most part) fixed by #11106. If there is still more we can do, a follow-up issue should be opened.

**Comment by bf** [comment]
Thank you for fixing this issue. It’s been seven years and I am glad that so many prominent names have interacted with my bug report.
